python/ray/data/tests/test_predicate_pushdown.py

```diff
@@ -369,6 +309,257 @@ def test_pushdown_with_rename_and_filter(
     pd.testing.assert_frame_equal(df, df1)
 
 
+def _get_optimized_plan(ds: Dataset) -> str:
+    """Get the optimized logical plan as a string."""
+    logical_plan = ds._plan._logical_plan
+    optimized_plan = LogicalOptimizer().optimize(logical_plan)
+    return optimized_plan.dag.dag_str
+
+
+def _check_plan_matches_pattern(ds: Dataset, expected_pattern: str):
+    """Check that the optimized plan matches the expected regex pattern."""
+    actual_plan = _get_optimized_plan(ds)
+    assert re.match(expected_pattern, actual_plan), (
+        f"Plan mismatch:\n"
+        f"Expected pattern: {expected_pattern}\n"
+        f"Actual plan: {actual_plan}"
+    )
+
+
+class TestPredicatePushdownIntoRead:
+    """Tests for pushing predicates into Read operators.
+
+    When a data source supports predicate pushdown (like Parquet),
+    the filter should be absorbed into the Read operator itself.
+    """
+
+    @pytest.fixture
+    def parquet_ds(self, ray_start_regular_shared):
+        return ray.data.read_parquet("example://iris.parquet")
+
+    def test_complex_pipeline_all_filters_push_to_read(self, parquet_ds):
+        """Complex pipeline: filters should push through all operators into Read.
+
+        Pipeline: Read -> Filter -> Rename -> Filter -> Sort -> Repartition
+                  -> Filter -> Limit -> Filter
+
+        All filters should fuse, push through all operators, rebind through rename,
+        and be absorbed into the Read operator.
+        """
+        ds = (
+            parquet_ds.filter(expr=col("sepal.length") > 4.0)
+            .rename_columns({"sepal.length": "len", "sepal.width": "width"})
+            .filter(expr=col("len") < 7.0)
+            .sort("len")
+            .repartition(3)
+            .filter(expr=col("width") > 2.5)
+            .limit(100)
+            .filter(expr=col("len") > 4.5)
+        )
+
+        # Verify correctness: should apply all filters correctly
+        expected = (
+            parquet_ds.filter(
+                expr=(col("sepal.length") > 4.0)
+                & (col("sepal.length") < 7.0)
+                & (col("sepal.width") > 2.5)
+                & (col("sepal.length") > 4.5)
+            )
+            .rename_columns({"sepal.length": "len", "sepal.width": "width"})
+            .sort("len")
+            .repartition(3)
+            .limit(100)
+        )
+
+        assert rows_same(ds.to_pandas(), expected.to_pandas())
+
+        # Verify plan: all filters pushed into Read, passthrough ops remain
+        plan = _get_optimized_plan(ds)
+        assert "Filter[Filter" not in plan, f"No Filter operators should remain: {plan}"
```

Comment:
Here and elsewhere -- would prefer to not test against the string representations because that's historically been pretty brittle.

If it'd make sense to do so, one option is to add a utility function to `LogicalPlan` to see it structurally matches another.